### DISCUSSION TITLE
Car detection fine tune

### DISCUSSION THREAD

**Discussion by fellipezambrano** [ultralytics/ultralytics#21473] [General]
Hello guys.

I'm working in a project which I need to count when cars crosses a line. Video source is a streaming from opencv2 working perfectly.

Some problems which I have found, and would like your support:
1-When car is moving frame by frame, Yolo box, with car id: In one video frame box is shown, with car ID. In the next frame, box disappears. This behavior is happening ramdonly. When car crosses the line, with the hidden box, car is not added to my car count. It is a problem to me, how to solve this?

2-As the problem 1, this also happens ramdonly. One car in the video with two boxes, with two IDs, then counting two cars when crossing the line. How to treat this?

Complementar information: I'm using YoloV11, S and M, but, not getting really good performance improvments, between S and M, with a improve of hardware consumption.

My code part with Yolo settings:
if cls_id in [2, 3, 5, 7] and conf > 0.01:
                        y_line = line_position[0][1]
                        if abs(center_y - y_line) < 5 and track_id not in tracked_ids:
                            car_count += 1
                            tracked_ids.add(track_id)

I set the conf very low looking for a way of reduce the first problem which car detection changes ramdonly in each frame, considering: 0,01, "car always will be there, in all frames", but not working as I hope and maybe causing the second problem.

Well, do you have please, any suggestion to help me?
Thanks a lot.
Best regards.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @fellipezambrano, thank you for sharing your project and questions with the Ultralytics community 🚗📈! This is an automated response to help get you started—an Ultralytics engineer will also assist you soon.

For new users, we recommend visiting the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions about detection, tracking, and count-based applications are covered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can better assist in debugging.

For custom training or fine-tuning questions, please include as much detail as possible, such as:
- Dataset image examples
- Training logs
- Details on your tracking approach and any third-party libraries used

Make sure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure optimal performance with YOLO11.

Join the Ultralytics community where it suits you best:
- For real-time chat, check out [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth technical discussions, visit [Discourse](https://community.ultralytics.com/)
- For sharing knowledge and experiences, join our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please ensure you are running the latest `ultralytics` package (which includes YOLO11) and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can run in any of the following up-to-date verified environments, with all dependencies (including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/)) preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience and for helping us improve Ultralytics!

**Reply by glenn-jocher** [reply]
Hi @fellipezambrano! Both issues you're experiencing are common tracking challenges. Setting confidence to 0.01 is actually causing more problems than solving them - this extremely low threshold creates noisy detections that confuse the tracker.

Try increasing your confidence to 0.3-0.5 and experiment with different tracking algorithms. You can specify the tracker with `model.track(source=video, tracker="bytetrack.yaml")` or `tracker="botsort.yaml"`. ByteTrack often performs better in scenarios with multiple similar objects like cars, while BoT-SORT handles erratic movement patterns well.

For the double detection issue, you might also want to adjust the IoU threshold for NMS by adding `iou=0.5` to your track parameters. If you're still getting inconsistent results, consider [custom training](https://docs.ultralytics.com/modes/train/) on your specific camera angle and lighting conditions, as this can significantly improve tracking stability for your use case.

The hardware difference between YOLO11s and YOLO11m might not be worth it if your tracking issues aren't resolved first - focus on getting stable detections before optimizing for speed.

**Reply by fellipezambrano** [reply]
Hello @glenn-jocher, Thanks for your kinddle support.

While I waited some answer here in forum, I kept doing my research. Copilot has been very useful for me. He suggested I use this file YAML, as you with extra configuration, which last test (I did just one - my solution need to count cars in a drive trhu of fast food, so I need to wait the costumers).

With the recomendations of copilot, I adjusted my YAML, as below:

tracker_type: botsort
with_reid: True
model: auto
gmc_method: sparseOptFlow
fuse_score: True
match_thresh: 0.6

track_high_thresh: 0.40
track_low_thresh: 0.01
new_track_thresh: 0.5
track_buffer: 600
proximity_thresh: 0.3
appearance_thresh: 0.6

Share with me, please your thoghs about that. As I said, last night it performed well, but was just the first test.

About making a training with Yolo, as you suggested, my qyestion is: How many samples of cars should I take to be better than the standart model of a car?! Maybe I'm wrong but, as a car is a very common object to detect, I think that it would be very hard, to my model getting better than the Yolo standart. M'I correct?

Again, thanks.
Best regards.

**Reply by glenn-jocher** [reply]
Your YAML configuration looks much better! The `track_high_thresh: 0.40` is a significant improvement over 0.01. However, I'd suggest raising `track_low_thresh` to around 0.15-0.20 and reducing `track_buffer` to 30-60 frames to prevent keeping dead tracks too long in a drive-thru scenario where cars move through quickly.

For custom training in your drive-thru use case, you're partially correct that cars are well-represented in standard datasets, but custom training can still be valuable for your specific conditions. You'd typically need 1000-2000 images minimum, but the real benefit comes from training on your exact camera angle, lighting conditions, and the specific types of vehicles that frequent your location. Since drive-thru cameras often have unique overhead angles and lighting challenges, custom training could improve consistency even with well-known objects like cars.

That said, try optimizing your tracking parameters first before investing time in custom training - your current approach with the improved YAML might solve your counting issues without the additional complexity.
